File: 1-intro/4-exploring-the-waters/cpp/arrayChange.cpp
int solution(vector<int> arr) {
    int moves = 0;
    for (int i = 0; i < arr.size() - 1; i++) {
        if (arr[i] >= arr[i + 1]) {
            const int diff = (arr[i] - arr[i + 1]) + 1;
            arr[i + 1] = arr[i] + 1;
            moves += diff;
        }
    }

    return moves;
}

/* -------------------------------------------------------------------------- */

int solution2(vector<int> arr) {
    int moves = 0;
    int m = -1e6;
    for (int val : arr) {
        m = max(m + 1, val);
        moves += m - val;
    }

    return moves;
}

File: 1-intro/4-exploring-the-waters/js/arrayChange.js
const solution = arr => {
  let moves = 0;
  for (let i = 0; i < arr.length - 1; i++) {
    if (arr[i] >= arr[i + 1]) {
      const diff = (arr[i] - arr[i + 1]) + 1;
      arr[i + 1] = arr[i] + 1;
      moves += diff;
    }
  }

  return moves;
}

/* -------------------------------------------------------------------------- */

const solution2 = arr => {
  const ans = arr.reduce((tot, val, idx, arr) => {
    while (val >= arr[idx + 1]) {
      tot++;
      val[idx + 1]++;
      return tot;
    }
  }, 0);

  return ans;
}
